Add rake email functionality to get incoming emails from Gmail API

Givo29 opened the pull request:
This PR adds on to the email rake functionality to allow users to pull emails from Gmail via the Gmail API. It was developed as we recently turned off all IMAP and POP3, ref: [ticket 31973](https://community.openproject.org/projects/openproject/work_packages/31973/activity) where we implemented incoming emails via IMAP and later created #8494, which was rejected.

## Setup
Using this feature does require some setup on Google Cloud, instructions on the setup have been attached at the bottom of the pull request as a pdf.

## Usage
Usage is the same as other incoming email rake commands:
`openproject run bundle exec rake redmine:email:receive_gmail <options>`

### Options
`credentials`: Gmail service account credentials file (JSON)
`username`: Gmail email address
`query`: Gmail search query (https://support.google.com/mail/answer/7190?hl=en)
`read_on_failure`: Mark emails as read even on failure (default: true)
`max_emails`: Max emails to process (default: 1000)


[Set Up Google Cloud For Gmail API Use.pdf](https://github.com/opf/openproject/files/10737454/Set.Up.Google.Cloud.For.Gmail.API.Use.pdf)

Givo29 committed:
Add rake email functionality to get incoming emails from Gmail API

Givo29 committed:
Fix default max_emails in help menu

oliverguenther commented on Gemfile:
```diff
@@ -311,6 +311,11 @@ gem 'roar', '~> 1.2.0'
+# Gmail API
+gem 'google-apis-core'
+gem 'google-apis-gmail_v1'
+gem 'googleauth'
```
```suggestion
gem 'google-apis-core', require: false
gem 'google-apis-gmail_v1', require: false
gem 'googleauth', require: false
```

As you require these dependencies in the rakefile, we can avoid loading them for the entire stack, which reduces memory consumption a bit.

oliverguenther commented on lib/redmine/gmail.rb:
```diff
@@ -0,0 +1,67 @@
+      end
+
+      def log_debug(&)
+        logger.debug(yield) if logger && logger.debug?
```
```suggestion
        logger.debug(&)
```

You can omit the check to debug?, as the block will only get evaluated if the log level is debug

oliverguenther reviewed (changes requested):
Hi @Givo29 , thanks for the contribution! This is very appreciated. I'll give this a try. I have two minor code changes.

It would be great if you could document this use-case in the inbound emails guide here: https://github.com/opf/openproject/blob/release/12.4/docs/installation-and-operations/configuration/incoming-emails/README.md

Givo29 committed:
Apply suggestions from code review

Don't require google api gems in Gemfile, omit redundant `logger.debug` check in gmail.rb

Co-authored-by: Oliver Günther <mail@oliverguenther.de>

Givo29 commented:
Thanks for the feedback @oliverguenther, what kind of information would you be looking for in terms of the documentation you requested exactly?

oliverguenther commented on Gemfile:
```diff
@@ -311,6 +311,11 @@ gem 'roar', '~> 1.2.0'
 gem 'rack-cors', '~> 1.1.1'
 
+# Gmail API
+gem 'google-apis-core', require: false
```
This line should be removed with `require: false`, otherwise bundler will complain as `google-apis-gmail_v1` will already require the gem (that's why the tests are failing)

oliverguenther replied:
@Givo29 this is still causing the tests to fail, simply remove that first require should allow the suite to pass.

Givo29 replied:
Ah sorry, I didn't see that comment. I will fix it now.

oliverguenther commented:
> Thanks for the feedback @oliverguenther, what kind of information would you be looking for in terms of the documentation you requested exactly?

I've seen you add a PDF with required steps, could you in this PR simply add some steps on how to set up the API key and call the rake task here: https://github.com/opf/openproject/blob/release/12.4/docs/installation-and-operations/configuration/incoming-emails/README.md#setup

Givo29 committed:
Add setup instructions and info about Gmail API Rake task to documentation

Givo29 committed:
Remove 'google-apis-core' causing tests to fail

oliverguenther commented:
I've rebased this to current dev in https://github.com/opf/openproject/pull/12231 and merged it. Thanks a lot @Givo29 👍
